# Lab 2: CS 6760


## Introduction to Privacy Engineering, Summer 2021


## Data Mining

##### Installation:

Same as Lab 1. <br>No additional packages are required.

##### This lab consists of 7 questions.

We will be using the 'Strava cycling data of multiple segments in Jeddah' dataset for this lab. 

##### Grading rubric: for each question,
100% of the points - Correct code, correct output<br>
50% of the points - Minor logical error, partial output<br>
0% of the points - No attempt, incomplete code, wrong output

##### Submission instructions

You will have to submit the completed jupyter notebook file (.ipynb) in Canvas.

### Note: The last cell of this .ipynb is a markdown cell for your answers. <br>Please fill this cell to complete your lab submission. 

The coding cells can be used for computations or observations. <br>


***
***

##### Dataset

##### Strava cycling data of multiple segments in Jeddah, Saudi Arabia

Strava is a social platform for cyclists and runners to share their activity and form groups.

A Strava Segment is an activity path with limited distance which anybody can take (given that they are recording an activity) to participate in a leaderboard to see who is the fastest.

Features:

- user_age_group: The age group of the participant
- user_id: The User ID of the participant
- attempt_date: The date of attempt of the entry in the leaderboard
- gender: The gender of the participant
- smt_rank: The participant's rank in the leaderboard
- smt_avg_spd: The participant's average speed in km/h within the segment
- smt_finish_seconds: The time taken for the participant to complete the segment in seconds
- smt_name: The name of the segment
- user_weight_category: The weight category of the participant
- act_title: The title of the activity which included the segment attempt
- act_avg_spd: The participant's average speed in km/h during the activity
- act_max_spd: The participant's maximum speed in km/h during the activity
- act_total_km: The total distance of the participant's activity in kilometers
- act_moving_seconds: The total time which the participant spent moving during the activity in seconds
- act_total_seconds: The total time of the activity (including stop times) in seconds
- has_hr_data: Whether there was data of participant's heart rate in the segment attempt

https://www.kaggle.com/baghlafturki/strava-jeddah-segments-leaderboard

This dataset is available in your HW zip file. 


***



#### Loading the file

In [1]:
# Loading libraries 

import numpy as np
import math
import pandas as pd

In [2]:
# Reading the data file
# Each row represents a user activity entry 
url = "https://raw.githubusercontent.com/icodealittle/SU2021_CS6760/main/jeddah_strava_segments.csv"
df = pd.read_csv(url)
df.head(5)

user_age_group  user_id  ... act_total_seconds has_hr_data
0   19 and under        1  ...              5147           0
1   19 and under        2  ...              4057           0
2   19 and under        3  ...              8962           0
3   19 and under        4  ...              5121           0
4   19 and under        5  ...             18504           0

[5 rows x 16 columns]

Each row of this dataframe represents a user activity entry. <br>
Each column of this dataframe represents a feature associated with the user entry. <br><br>

In [3]:
# Printing the rows and columns

print("The dataframe contains ",df.shape[0]," rows.")
print("The dataframe contains ",df.shape[1]," columns.")

The dataframe contains  7865  rows.
The dataframe contains  16  columns.


In [4]:
# Printing the names of the columns

for col in df.columns: 
    print(col) 

user_age_group
user_id
attempt_date
gender
smt_rank
smt_avg_spd
smt_finish_seconds
smt_name
user_weight_category
act_title
act_avg_spd
act_max_spd
act_total_km
act_moving_seconds
act_total_seconds
has_hr_data


The description of these features can be found at the first markdown cell ( on top ) of this .ipynb
<br><br>
Among these columns, let's look at the user age group.

In [5]:
# The filter function is used here to extract a subset of columns from the dataset.

df.filter(['user_age_group'])

user_age_group
0      19 and under
1      19 and under
2      19 and under
3      19 and under
4      19 and under
...             ...
7860       70 to 74
7861            75+
7862            75+
7863            75+
7864            75+

[7865 rows x 1 columns]

What are the user age groups  present in this dataset? <br><br>
We can use Pandas' "unique" function to filter out all the repeating values in a column. <br>

In [6]:
# Printing the unique values 
for grp in df.user_age_group.unique():
    print(grp)

19 and under
20 to 24
25 to 34
35 to 44
45 to 54
55 to 64
65 to 69
75+
70 to 74


How many user entries are present per age group? 
 
<br>Let's find out the number of user entries per age group.
<br> Since each row represents a user entry, <br>
We want to count the number of rows for each of the 'user_age_group' categories. 

<br> Using Pandas,
<br> We can do this by grouping the data based on the column values in 'user_age_group' <br> and counting the number of rows for each group.
<br> The functions used are 'groupby' and 'count'.

In [7]:
# Using grouping and counting to find the number of user entries per age group.

df.groupby('user_age_group').count()


user_id  attempt_date  ...  act_total_seconds  has_hr_data
user_age_group                         ...                                
19 and under        370           370  ...                370          370
20 to 24            705           705  ...                705          705
25 to 34           3353          3353  ...               3353         3353
35 to 44           2323          2323  ...               2323         2323
45 to 54            880           880  ...                880          880
55 to 64            192           192  ...                192          192
65 to 69             21            21  ...                 21           21
70 to 74              3             3  ...                  3            3
75+                  18            18  ...                 18           18

[9 rows x 15 columns]

We can observe that the row counts have been computed for each category.

#### Question 1 ( 3 points )

##### Which named segment has the least number of user activity entries?

Named segments are the named path segments for the cycling/activity routes.
<br> Identify the column that contains this information
<br> Use this column to find out the number (count) of user activity entries (rows) for each named segment. 
<br> From these counts, find out which named segment has the least count.

In [8]:
## You can use this cell for Question 1
## YOUR CODE HERE:

least_path = df['smt_name'].value_counts().idxmin()

#Test
print(least_path)

Al Fardoos to shellfish round about


#### Question 2 ( 3 points )

##### Which user weight category has the most number of user activity entries?
<br> Similar to Q.1

In [9]:
## You can use this cell for Question 2
## YOUR CODE HERE:

most_weight = df['user_weight_category'].value_counts().idxmax()

#Test
print(most_weight)

54 kg and under


#### Question 3 ( 3 points )

##### Which numerical feature has the highest magnitude of correlation with 'smt_rank' ?
'smt_rank' refers to the participant's rank in the leaderboard

The correlation of a variable with itself is always 1. <br>
We are looking for a numerical feature other than 'smt_rank' itself. 
<br> For this question, the correlation matrix is provided to you as a dataframe.

In [10]:
#Standard correlation for the numerical features in this dataframe.

df_corr = df.select_dtypes(['number']).corr(method='pearson')
df_corr 

user_id  smt_rank  ...  act_total_seconds  has_hr_data
user_id             1.000000  0.273225  ...           0.008099    -0.091608
smt_rank            0.273225  1.000000  ...           0.017735    -0.220207
smt_avg_spd        -0.353661 -0.546382  ...          -0.025865     0.280936
smt_finish_seconds -0.014593  0.000682  ...           0.012026    -0.029537
act_avg_spd        -0.371738 -0.494412  ...          -0.106255     0.387707
act_max_spd        -0.125751 -0.136210  ...           0.053858     0.005203
act_total_km       -0.214994 -0.285394  ...           0.036902     0.176928
act_moving_seconds  0.008099  0.017735  ...           1.000000    -0.013010
act_total_seconds   0.008099  0.017735  ...           1.000000    -0.013010
has_hr_data        -0.091608 -0.220207  ...          -0.013010     1.000000

[10 rows x 10 columns]

In [11]:
## You can use this cell for Question 3
## YOUR CODE HERE:
correlation_sorting = df_corr['smt_rank'].abs().sort_values(ascending = False)

best_name = correlation_sorting.index[1]
best_value = correlation_sorting[1]

#Test
print(best_name, best_value)

smt_avg_spd 0.5463821107460634


#### Question 4 ( 5 points)

The unique identifiers ( values with frequency = 1 ) present in the columns/features of this dataset can be used to deterministically identify a user segment entry or a user ID. <br>The unique identifiers can result in de-anonymization of the data entry or the user. 

##### Among the 16 features/columns of this dataset, which feature(s) have the highest number of unique identifiers associated with the row (user entry) ?
##### How many unique identifiers do these column(s) have?

To find the number of unique identifiers,
<br> For each column,
<br> We identify the values that have frequency (row-count) = 1
<br> These are the unique identifiers present in that column
<br> Count these unique identifiers for each column. 
<br> If this is done, For each column, we have the number of unique identifiers.
<br> The column(s) with the highest number of unique identifiers can now be determined.
 


In [12]:
## You can use this cell for Question 4
## YOUR CODE HERE:

df.nunique().sort_values(ascending=False)   

act_total_seconds       4275
act_moving_seconds      4275
act_total_km            3805
user_id                 2254
attempt_date            1528
smt_finish_seconds      1301
act_title               1213
smt_rank                 735
smt_avg_spd              475
act_avg_spd              338
act_max_spd              283
smt_name                   9
user_age_group             9
user_weight_category       8
has_hr_data                2
gender                     2
dtype: int64

#### Question 5 ( 3 points )

Binning or discretization can reduce the risk of de-anonymization in a dataset. 

##### If the 'act_moving_seconds' is replaced with the nearest integer-valued minute, <br> what is the modified number of unique identifiers present in this column? 
Steps:
- replace 'act_moving_seconds' with the nearest integer-valued minute.
- find the new/modified number of unique identifiers for this column ( same steps as Q.4 )

<br> To replace the values in a column, take a look at this following example:

In [24]:
## Example for replacing values in a column
## In this example, we are replacing the user_ID with 42 x user_ID

## We don't want to disturb the original dataframe.
df_example = df.copy()

print(df_example.filter(['user_id']).head(5))
print("\n \n Example: replacing the user_ID with 10 x user_ID  \n \n")

## Pay attention to this: 
df_example['user_id'] = df_example.apply (lambda row: row['user_id']*42 , axis=1)

print(df_example.filter(['user_id']).head(5))

   user_id
0        1
1        2
2        3
3        4
4        5

 
 Example: replacing the user_ID with 10 x user_ID  
 

   user_id
0       42
1       84
2      126
3      168
4      210


In [27]:
## You can use this cell for Question 5
## YOUR CODE HERE:
df_unique = df.copy()
df_unique['act_total_seconds'] /= 60
df_unique['act_total_seconds'] = df_unique['act_total_seconds'].astype(int)

newUniqueNumber = df_unique.nunique().sort_values(ascending=False)
newMinutesAct = newUniqueNumber['act_total_seconds']

#Test
print('act_total_minutes', newMinutesAct)

#From Optional OH on Sunday
# df_example['act_moving_seconds'] = df_example.apply(lambda row: int(row['act_moving_seconds']/60), axis = 1)
# print(df_example.filter(['act_moving_seconds']))

# df_act = df_example.columns

# for e in df_act:
#   unique_ID = df_example.groupby(e).count()
#   unique_ID = unique_ID.loc[unique_ID.iloc[:,0] ==1]
#   count = unique_ID.shape[0]
#   print(e, count)


act_total_minutes 363


You will notice that for this column ( 'act_moving_seconds' ), the modified number of unique identifiers is significantly lower than the number of unique identifiers before modification.

#### Question 6 ( 4 points )

Consider the information present in these 3 columns : <br>'user_id', 'user_age_group' and 'gender'


In [15]:
df6 = df.filter(['user_id','user_age_group','gender'])
df6.head(10)

user_id user_age_group gender
0        1   19 and under   male
1        2   19 and under   male
2        3   19 and under   male
3        4   19 and under   male
4        5   19 and under   male
5        6   19 and under   male
6        7   19 and under   male
7        8   19 and under   male
8        9   19 and under   male
9       10   19 and under   male

Given only this data subset, among the different 'gender'-'user-age-group' pairings in this subset,

#####  Which pairing poses the highest risk of de-anonymization for the associated users and their 'user_id'?
##### Which pairing poses the least risk?

Pairing: refers to all the possible 'gender'-'user-age-group' pairs in this dataset. 
<br> Hint: Pandas grouping can be done using more than 1 column. <br>
<br> Risk: For this specific question, building from (Q.4), risk refers to the amount of data that can be de-anonymized using unique identifiers. <br>
<br> If a pairing has more unique identifiers, the users associated with that pairing have higher risk of being de-anonymized.

In [16]:
## You can use this cell for Question 6
## YOUR CODE HERE:

gender_age = df6.groupby(['gender','user_age_group']).count()
highly_likely_risk = gender_age['user_id'].idxmin()
less_likely_risk = gender_age['user_id'].idxmax()

#Test
print('The group that most likely to Risk = ', highly_likely_risk)
print('The group that least likely to Risk = ', less_likely_risk)


The group that most likely to Risk =  ('female', '65 to 69')
The group that least likely to Risk =  ('male', '25 to 34')


#### Question 7 ( 4 points )

For the dataframe ( 'df6' containing 3 features )  in Question 6, 


##### Recommend a method to ensure that all the users associated with the different 'gender'-'user-age-group' pairings are equally protected. 
( expected: 2-3 sentences ) 

##### Does your recommendation contribute to any additional risk of de-anonymization?   
( expected: 2-3 sentences)

<br> Hint: Observing the counts of unique identifiers associated with the pairings (Q.6) can be helpful. 

Enter your answer in the 'Submitted answers' cell.

***

( Double click the cell below to edit it )

***

# Submitted answers:


* Q.1 : [3] Least number of user activity entries - <b>Al Fardoos to shellfish round about</b>


* Q.2 : [3] Weight category with most entries - <b>54 kg and under</b>


* Q.3 : [3] Feature with highest magnitude of correlation - <b>smt_avg_spd 0.5463821107460634</b>


* Q.4 : 
   * (a) [3] Column names -  <b>

   act_total_seconds
   
   act_moving_seconds      
   act_total_km            
   user_id                 
   attempt_date            
   smt_finish_seconds      
act_title               
smt_rank                 
smt_avg_spd              
act_avg_spd              
act_max_spd              
smt_name                  
user_age_group             
user_weight_category       
has_hr_data                
gender                     </b>

   * (b) [2] No. of unique identifiers - <b>       4275,
     4275,
     3805,              2254,
     1528,
      1301,
      1213,
      735,
   475,       338,
   283,         9,          9,   8,  2,        2</b>


* Q.5 [3] : Modified number of unique identifiers - <b>aact_total_minutes 363</b>


* Q.6 : 
   * (a) [2] Pairing with highest risk -  <b>'female', '65 to 69'</b>

   * (b) [2] Pairing with lowest risk - <b>'male', '25 to 34'</b>


* Q.7 :
   * (a) [2] Recommendation -  <b> One method can help to ensure that all the users associated with the different 'gender' and 'user_age_group' is implementing the redistribution of the age grouping. By implementing this method, we can see the majority of the user within the dataset would fall in the age group of 25 to 44 age range. By breaking the group into smaller age group, it easier to analyst the group more efficiently.</b>

   * (b) [2] Additional risk - <b>he recommendation would creates additional risk by having too many granual information for the dateset. By having a large group of indivuals in 25 to 44 age range, it would protect them from being identify. Though if we create and break the group down, individuals within the rnage would have a higher risk of being identified.</b>
